core/server/master/src/main/java/alluxio/master/file/DefaultFileSystemMaster.java

```diff
@@ -3226,55 +2885,51 @@ private void setUfsAcl(RpcContext rpcContext, LockedInodePath inodePath)
         try {
           ufs.setAcl(ufsUri, inode.getACL());
           if (inode.isDirectory()) {
-            InodeDirectory inodeDirectory = (InodeDirectory) inode;
+            InodeDirectoryView inodeDirectory = (InodeDirectoryView) inode;
             ufs.setAcl(ufsUri, inodeDirectory.getDefaultACL());
           }
         } catch (IOException e) {
           throw new AccessControlException("Could not setAcl for UFS file: " + ufsUri);
         }
-
       }
     }
   }
 
   private void setAclSingleInode(RpcContext rpcContext, SetAclAction action,
       LockedInodePath inodePath, List<AclEntry> entries, boolean replay, long opTimeMs)
       throws IOException, FileDoesNotExistException {
-    Inode<?> targetInode = inodePath.getInode();
-    switch (action) {
-      case REPLACE:
-        // fully replace the acl for the path
-        targetInode.replaceAcl(entries);
-        break;
-      case MODIFY:
-        targetInode.setAcl(entries);
-        break;
-      case REMOVE:
-        targetInode.removeAcl(entries);
-        break;
-      case REMOVE_ALL:
-        targetInode.removeExtendedAcl();
-        break;
-      case REMOVE_DEFAULT:
-        targetInode.setDefaultACL(new DefaultAccessControlList(targetInode.getACL()));
-        break;
-      default:
+    InodeView inode = inodePath.getInode();
+
+    // Check that we are not removing an extended mask.
+    if (action == SetAclAction.REMOVE) {
+      for (AclEntry entry : entries) {
+        if (entry.getType() == AclEntryType.MASK) {
```

Comment:
I think this code disallows removing of all masks? we need to check if the ACL is extended.

Reply:
Good catch, updated to check `inode.getACL().hasExtended()` as well